Android/AppsFlyerURLResolver/AppsFlyerURLResolverLib/src/main/java/com/appsflyer/resolver/URLResolver.kt

```diff
@@ -1,66 +1,96 @@
-                    add(url)
-                }
+        startResolvingJob(url, urlResolverListener) {
+            it?.run {
+                var redirect = this
                 var res: AFHttpResponse? = null
                 for (i in 0 until maxRedirections) {
                     // resolve current URL - check for redirection
-                    res = resolveInternal(redirects.last())
-                    res.redirected?.let { // if redirected to another URL
-                        redirects.add(it)
-                    } ?: break
+                    res = resolveInternal(redirect) { con ->
+                        if (con.isRedirected) {
+                            // redirect
+                            con.redirectedTo.also { l ->
+                                afDebugLog("redirecting to $l")
+                            }
+                        } else {
+                            // did not redirected
+                            null
+                        }
+                    }
+                    res.redirected?.let { r ->
+                        redirect = r // if redirected to another URL - update last url
+                    } ?: break // break will not work inside `also` block on `resolveInternal`
                 }
-                if (res?.error == null) {
-                    afDebugLog("found link: ${redirects.last()}")
-                    urlResolverListener.onComplete(redirects.last())
-                } else {
-                    urlResolverListener.onComplete(null)
+                res?.apply {
+                    // return the last link we found
+                    redirected = redirect
+                }
+            }
+        }
+    }
+
+    fun resolveJSRedirection(url: String?, urlResolverListener: URLResolverListener) {
+        startResolvingJob(url, urlResolverListener) {
+            it?.run {
+                resolveInternal(this) { con ->
+                    if (con.isSuccessful && con.isHtmlPage) {
+                        con.jsRedirected
+                    } else {
+                        null
+                    }
                 }
-            } else {
-                urlResolverListener.onComplete(url)
             }
-        }.start()
+        }
     }
 
-    private fun resolveInternal(uri: String): AFHttpResponse {
+    private fun startResolvingJob(
+        url: String?,
+        urlResolverListener: URLResolverListener,
+        job: ResolveJob
+    ) {
+        Executors.newSingleThreadExecutor().submit {
+            url?.let {
+                afDebugLog("resolving $it")
+                if (it.isValidURL) {
+                    job(it)?.apply {
+                        error?.let {
+                            urlResolverListener.onComplete(null)
+                        } ?: redirected.run {
+                            afDebugLog("found link: $this")
+                            urlResolverListener.onComplete(this)
```

Comment:
The callbacks should be fired in the main thread not in background one

Reply:
done